llvm/lib/Target/PowerPC/PPCInstrPaired.td

```diff
@@ -0,0 +1,302 @@
+      def PSQ_STX : PSQForm_xw<7, (outs), (ins psrc:$fD, memrr:$dst, u1imm:$W, u3imm:$I),
+                              "psq_stx $fD, $dst, $W, $I", IIC_LdStStore, []>;
+
+      def PSQ_STUX : PSQForm_xw<39, (outs ptr_rc_nor0:$ea_result),
+                                (ins psrc:$fD, memrr:$dst, u1imm:$W, u3imm:$I),
+                                "psq_stux $fD, $dst, $W, $I", IIC_LdStSTUX, []>,
+                                RegConstraint<"$dst.ptrreg = $ea_result">,
+                                NoEncode<"$ea_result">;
+      let isCodeGenOnly = 1 in {
+        def PS_STX : PSNQForm_xw<7, 0, (outs), (ins psrc:$fD, memrr:$dst),
+                                 "ps_stx $fD, $dst", IIC_LdStStore,
+                                 [(int_ppc_paired_st xoaddr:$dst, v2f32:$fD)]>;
+        def PS_STWX : PSNQForm_xw<7, 1, (outs), (ins psrc:$fD, memrr:$dst),
+                                 "ps_stwx $fD, $dst", IIC_LdStStore,
+                                 [(int_ppc_paired_stw xoaddr:$dst, v2f32:$fD)]>;
+        def PS_STUX : PSNQForm_xw<39, 0, (outs ptr_rc_nor0:$ea_result),
+                                  (ins psrc:$fD, memrr:$dst),
+                                  "ps_stux $fD, $dst", IIC_LdStSTUX, []>,
+                                  RegConstraint<"$dst.ptrreg = $ea_result">,
+                                  NoEncode<"$ea_result">;
+        def PS_STWUX : PSNQForm_xw<39, 1, (outs ptr_rc_nor0:$ea_result),
+                                   (ins psrc:$fD, memrr:$dst),
+                                   "ps_stwux $fD, $dst", IIC_LdStSTUX, []>,
+                                   RegConstraint<"$dst.ptrreg = $ea_result">,
+                                   NoEncode<"$ea_result">;
+      }
+    }
+
+    defm PS_ABS : PSForm_xd0br<264, (outs psrc:$fD), (ins psrc:$fB),
+                               "ps_abs", "$fD, $fB", IIC_FPGeneral,
+                               [(set v2f32:$fD, (fabs v2f32:$fB))]>;
+
+    let isCommutable = 1 in {
+      defm PS_ADD : PSForm_adab0r<21, (outs psrc:$fD), (ins psrc:$fA, psrc:$fB),
+                                  "ps_add", "$fD, $fA, $fB", IIC_FPAddSub,
+                                  [(set v2f32:$fD, (fadd v2f32:$fA, v2f32:$fB))]>;
+    }
+
+    let isCompare = 1, mayRaiseFPException = 1, hasSideEffects = 0 in {
+      def PS_CMPU0 : PSForm_xcrab<0, 0, (outs crrc:$crD), (ins psrc:$fA, psrc:$fB),
+                                  "ps_cmpu0 $crD, $fA, $fB", IIC_FPCompare, []>;
+      def PS_CMPU1 : PSForm_xcrab<0, 1, (outs crrc:$crD), (ins psrc:$fA, psrc:$fB),
+                                  "ps_cmpu1 $crD, $fA, $fB", IIC_FPCompare, []>;
+      def PS_CMPO0 : PSForm_xcrab<1, 0, (outs crrc:$crD), (ins psrc:$fA, psrc:$fB),
+                                  "ps_cmpo0 $crD, $fA, $fB", IIC_FPCompare, []>;
+      def PS_CMPO1 : PSForm_xcrab<1, 1, (outs crrc:$crD), (ins psrc:$fA, psrc:$fB),
+                                  "ps_cmpo1 $crD, $fA, $fB", IIC_FPCompare, []>;
+    }
```

Comment:
Find out what the best way is to codegen these instructions. `ps_cmpu0` operates like `cmpu` on the 0th element of the vectors passed in, etc.